### PR TITLE
Implement GRAPHITE Attack

### PR DESCRIPTION
**Pull request by ryan-feng**
# Description

Add GRAPHITE attack (https://arxiv.org/pdf/2002.07088.pdf, https://github.com/ryan-feng/GRAPHITE)

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [X] Tested with notebooks
- [X] Independently tested for correctness
- [ ] Test cases added, did not know how to run outside of the PR

**Test Configuration**:
- OS - Ubuntu
- Python 3.8
- ART version or commit number: f073851ae325a3ac50d9f53e54f2cf14f9b95781
- Torch version 1.7

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 1]
```diff
@@ -0,0 +1,889 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2019
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:3]: Change the year to 2022

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+            x_adv = np.transpose(x_adv, (0, 3, 1, 2))
+
+        return x_adv
+
+    def _perturb(
+        self,
+        x: np.ndarray,
+        y: int,
+        x_tar: np.ndarray,
+        mask: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:310]: I think if you are using optional here, you should explicitly set the default value to None. You might also want to move it to the end of the arg list. This also occurs on lines 311, 373, 502, 547, 593, 656, 731,

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+
+    def _eval(
+        self,
+        x: "torch.Tensor",
+        x_adv: "torch.Tensor",
+        mask: "torch.Tensor",
+        target_label: int,
+        y_onehot: "torch.Tensor",
+        xforms: List[Tuple[float, float, float, int, float, float, float, float, float]],
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:295]: Should assign it with a default value of None and move to the end of the arg list.

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+                success = int(logits.argmax(dim=1).detach().cpu().numpy()[0] == target_label)
+                successes += success
+        return successes / len(xforms)
+
+    def _perturb(
+        self,
+        x: np.ndarray,
+        y: np.ndarray,
+        mask: np.ndarray,
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:332]: Should assign it with a default value of None and move to the end of the arg list.

File: art/attacks/evasion/graphite/utils.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2019
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:3]: Changed to 2022

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,565 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2019
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:3]: change to 2022

File: art/attacks/evasion/graphite/utils.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+    return images
+
+
+def transform_wb(
+    x: "torch.Tensor",
+    x_adv: "torch.Tensor",
+    mask: "torch.Tensor",
+    xform: Tuple[float, float, float, int, float, float, float, float, float],
+    pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:303]: Should assign it with a default value of None and move to the end of the arg list.

File: art/attacks/evasion/graphite/utils.py [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+    img: Union[np.ndarray, "torch.Tensor"],
+    angle: float,
+    width: int,
+    height: int,
+    focal: float,
+    dist: float,
+    crop_percent: float,
+    crop_off_x: float,
+    crop_off_y: float,
+    pts: np.ndarray,
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:406]: Why is this not marked as optional? On line 355, it gets passed pts, which could be optional.

File: art/attacks/evasion/graphite/utils.py [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+
+
+def get_offset_and_crop_size(
+    width: int,
+    height: int,
+    h_mat: np.ndarray,
+    crop_percent: float,
+    crop_off_x: float,
+    crop_off_y: float,
+    pts: np.ndarray,
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:496]: I think on line 355, this arg could be optional, but it is not marked as optional here.

File: tests/attacks/test_graphite.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2018
```
Comment (keykholt) [inline tests/attacks/test_graphite.py:3]: Change to 2022

**Review by keykholt — COMMENTED** [review]
Made some minor comments about type hinting.

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py (hunk 6 above)
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: art/attacks/evasion/graphite/graphite_blackbox.py (hunk 1 above)
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: art/attacks/evasion/graphite/graphite_blackbox.py (hunk 2 above)
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:310]:
```suggestion
        mask: Optional[np.ndarray] = None,
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 11]
```diff
@@ -0,0 +1,889 @@
...
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:311]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_blackbox.py:311]: Move to end of arg list as well

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 12]
```diff
@@ -0,0 +1,889 @@
...
+
+    def _generate_mask(
+        self,
+        x: np.ndarray,
+        x_noise: np.ndarray,
+        x_tar: np.ndarray,
+        x_tar_noise: np.ndarray,
+        mask: np.ndarray,
+        y: int,
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:373]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_blackbox.py:373]: Move to end of arg list

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 13]
```diff
@@ -0,0 +1,889 @@
...
+        return best_mask
+
+    def _get_heatmap(
+        self,
+        x: np.ndarray,
+        x_noise: np.ndarray,
+        x_tar_noise: np.ndarray,
+        mask: np.ndarray,
+        y: int,
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:502]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_blackbox.py:502]: Move to end of arg list also

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 14]
```diff
@@ -0,0 +1,889 @@
...
+    def _evaluate_transform_robustness_at_pivot(
+        self,
+        x: np.ndarray,
+        x_noise: np.ndarray,
+        x_tar_noise: np.ndarray,
+        y: int,
+        mask: np.ndarray,
+        patches: List[np.ndarray],
+        xforms: List,
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:547]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_blackbox.py:547]: Move to end of arg list

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 15]
```diff
@@ -0,0 +1,889 @@
...
+        self,
+        x: np.ndarray,
+        x_noise: np.ndarray,
+        x_tar_noise: np.ndarray,
+        y: int,
+        mask: np.ndarray,
+        patches: List[np.ndarray],
+        indices: List[Tuple[int, int]],
+        xforms: List,
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:593]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_blackbox.py:593]: Move to end of arg list

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 16]
```diff
@@ -0,0 +1,889 @@
...
+    def _get_fine_reduced_mask(
+        self,
+        x: np.ndarray,
+        x_noise: np.ndarray,
+        x_tar_noise: np.ndarray,
+        y: int,
+        mask: np.ndarray,
+        patches: List[np.ndarray],
+        xforms: List,
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:656]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_blackbox.py:656]: Move to end of arg list

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 17]
```diff
@@ -0,0 +1,889 @@
...
+        return best_mask
+
+    def _boost(
+        self,
+        x: np.ndarray,
+        x_noise: np.ndarray,
+        x_tar_noise: np.ndarray,
+        mask: np.ndarray,
+        y: int,
+        pts: Optional[np.ndarray],
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:731]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_blackbox.py:731]: Move to end of arg list

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py (hunk 3 above)
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:295]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:295]: Move to end of arg list

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py (hunk 4 above)
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:332]:
```suggestion
        pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [reply art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:332]: Move to end of arg list also

File: art/attacks/evasion/graphite/utils.py (hunk 5 above)
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: art/attacks/evasion/graphite/utils.py (hunk 7 above)
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:303]:
```suggestion
    pts: Optional[np.ndarray] = None,
```

Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:303]: This is tagged as optional, but then passed to other functions that mark it as required?

**Review by keykholt — CHANGES_REQUESTED** [review]

File: art/attacks/evasion/graphite/utils.py [hunk 18]
```diff
@@ -0,0 +1,627 @@
...
+        max_y = min(img.shape[0], img_tensor.shape[1])
+        max_x = min(img.shape[1], img_tensor.shape[2])
+        img_tensor[count, :max_y, :max_x, :] = img[:max_y, :max_x, :]
+        count += 1
+        if count == batch_size or i == len(imgs) - 1:
+            if count < batch_size:
+                img_tensor = img_tensor[:count, :, :, :]
+                tar_tensor = tar_tensor[:count]
+
+            if estimator.channels_first:
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:601]: I think you might want to use a different type or check to make sure that channels_first is an attribute

File: art/attacks/evasion/graphite/utils.py [hunk 19]
```diff
@@ -0,0 +1,634 @@
...
+        dist,
+        crop_percent,
+        crop_off_x,
+        crop_off_y,
+        pts,
+        whitebox=True,
+    )
+
+    # make mypy happy
+    assert isinstance(x_adv_tmp, "torch.Tensor")
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:369]: Probably wouldn't have an assert here. Also, I think the string --> tensor will only work if type checking is enabled. Also, can't get_perspective_transform also return a numpy type>

If you do want an assert, replace it with an `if isinstance: raise SpecificError(errormsg)` form

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 20]
```diff
@@ -0,0 +1,579 @@
...
+            # make mypy happy
+            assert isinstance(transform_robustness, float)
+            if transform_robustness < self.min_tr:
+                break
+
+            prev_attack = adv_img.detach().clone()
+
+            # Remove low-impact pixel-patches or pixels in mask.
+            # make mypy happy
+            assert isinstance(final_avg_grad, "torch.Tensor")
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:465]: Can't use a string here as it only turns into a tensor type if type checking is being done I think. Also, if you want to type check, you should use an `if  not isinstance(): raise SpecificError(errormsg)`.

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 21]
```diff
@@ -0,0 +1,579 @@
...
+                    pts,
+                    clip_min,
+                    clip_max,
+                )
+                final_avg_grad = avg_grad
+                if transform_robustness >= self.min_tr:
+                    break
+
+            # make mypy happy
+            assert isinstance(transform_robustness, float)
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:457]: if you want to type check, you should use an if not isinstance(): raise SpecificError(errormsg).

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 22]
```diff
@@ -0,0 +1,894 @@
...
+        mask = kwargs.get("mask")
+        x_tar = kwargs.get("x_tar")
+        obj_width = kwargs.get("obj_width") if "obj_width" in kwargs else 30
+        focal = kwargs.get("focal") if "focal" in kwargs else 3
+        pts = kwargs.get("pts") if "pts" in kwargs else None
+
+        if y is None:
+            raise ValueError("Target labels `y` need to be provided.")
+
+        assert isinstance(y, np.ndarray)
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:212]: if you want to type check, you should use an if not isinstance(): raise SpecificError(errormsg).

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 23]
```diff
@@ -0,0 +1,894 @@
...
+        x_tar = (x_tar.copy() - clip_min) / (clip_max - clip_min)
+
+        if mask is None:
+            mask = np.ones((self.noise_size[1], self.noise_size[0], x.shape[2]))
+        mask = mask / np.max(mask)
+
+        x_copy = x.copy()
+        x_tar_copy = x_tar.copy()
+        # make mypy happy
+        assert isinstance(mask, np.ndarray)
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:349]: if you want to type check, you should use an if not isinstance(): raise SpecificError(errormsg).

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 24]
```diff
@@ -0,0 +1,887 @@
...
+        :param pts: Optional. A set of points that will set the crop size in the perspective transform.
+        :return: An adversarial example.
+        """
+        import cv2
+
+        x = (x.copy() - clip_min) / (clip_max - clip_min)
+        x_tar = (x_tar.copy() - clip_min) / (clip_max - clip_min)
+
+        if mask is None:
+            mask = np.ones((self.noise_size[1], self.noise_size[0], x.shape[2]))
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:338]:
```suggestion
        if mask is None:
            mask_array = np.ones((self.noise_size[1], self.noise_size[0], x.shape[2]))
        else:
            mask_array = mask
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 25]
```diff
@@ -0,0 +1,887 @@
...
+        mask = mask / np.max(mask)
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:339]: Renaming to the new variable name
```suggestion
        mask_array = mask_array / np.max(mask_array)
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 26]
```diff
@@ -0,0 +1,889 @@
...
+        y: int,
+        obj_width: float,
+        focal: float,
+        clip_min: float,
+        clip_max: float,
+        pts: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        """
+        Function to generate a mask.
+        :param x: An array with one original input to be attacked.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:382]:
```suggestion

        :param x: An array with one original input to be attacked.
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 27]
```diff
@@ -0,0 +1,889 @@
...
+        y: int,
+        patches: List[np.ndarray],
+        xforms: List,
+        clip_min: float,
+        clip_max: float,
+        pts: Optional[np.ndarray] = None,
+    ) -> List[float]:
+        """
+        Function to generate a heatmap.
+        :param x: An array with one original input to be attacked.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:510]:
```suggestion

        :param x: An array with one original input to be attacked.
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 28]
```diff
@@ -0,0 +1,889 @@
...
+        patches: List[np.ndarray],
+        xforms: List,
+        clip_min: float,
+        clip_max: float,
+        pivot: int,
+        pts: Optional[np.ndarray] = None,
+    ) -> Tuple[float, np.ndarray]:
+        """
+        Function as a binary search plug-in that evaluates the transform-robustness at the specified pivot.
+        :param x: An array with one original input to be attacked.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:554]:
```suggestion

        :param x: An array with one original input to be attacked.
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 29]
```diff
@@ -0,0 +1,889 @@
...
+        patches: List[np.ndarray],
+        indices: List[Tuple[int, int]],
+        xforms: List,
+        clip_min: float,
+        clip_max: float,
+        pts: Optional[np.ndarray] = None,
+    ) -> Tuple[np.ndarray, List[np.ndarray], List[Tuple[int, int]]]:
+        """
+        Function to coarsely reduce mask.
+        :param x: An array with one original input to be attacked.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:599]:
```suggestion

        :param x: An array with one original input to be attacked.
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 30]
```diff
@@ -0,0 +1,889 @@
...
+        xforms: List,
+        object_size: float,
+        clip_min: float,
+        clip_max: float,
+        lbd: float = 5,
+        pts: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        """
+        Function to finely reduce mask.
+        :param x: An array with one original input to be attacked.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:664]:
```suggestion

        :param x: An array with one original input to be attacked.
```

File: art/attacks/evasion/graphite/graphite_blackbox.py [hunk 31]
```diff
@@ -0,0 +1,889 @@
...
+        y: int,
+        obj_width: float,
+        focal: float,
+        clip_min: float,
+        clip_max: float,
+        pts: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        """
+        Function to boost transform-robustness.
+        :param x: An array with one original input to be attacked.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_blackbox.py:739]:
```suggestion

        :param x: An array with one original input to be attacked.
```

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 32]
```diff
@@ -0,0 +1,575 @@
...
+        target_label: np.ndarray,
+        y_onehot: "torch.Tensor",
+        xforms: List[Tuple[float, float, float, int, float, float, float, float, float]],
+        pts: Optional[np.ndarray],
+        clip_min: float,
+        clip_max: float,
+    ) -> float:
+        """
+        Compute transform-robustness.
+        :param x: Original image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:301]:
```suggestion

        :param x: Original image.
```

File: art/attacks/evasion/graphite/graphite_whitebox_pytorch.py [hunk 33]
```diff
@@ -0,0 +1,575 @@
...
+        mask: np.ndarray,
+        pts: Optional[np.ndarray],
+        obj_width: float,
+        focal: float,
+        clip_min: float,
+        clip_max: float,
+    ) -> np.ndarray:
+        """
+        Internal attack function for one example.
+        :param x: An array with one original input to be attacked.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/graphite_whitebox_pytorch.py:340]:
```suggestion

        :param x: An array with one original input to be attacked.
```

File: art/attacks/evasion/graphite/utils.py [hunk 34]
```diff
@@ -0,0 +1,671 @@
...
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+    import torch
+
+_estimator_requirements = (BaseEstimator, ClassifierMixin)
+
+
+def dist2pixels(dist: float, width: float, obj_width: float = 30) -> float:
+    """
+    Convert distance to pixels.
+    :param dist: Distance to object.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:64]:
```suggestion

    :param dist: Distance to object.
```

File: art/attacks/evasion/graphite/utils.py [hunk 35]
```diff
@@ -0,0 +1,671 @@
...
+    :return: Distance in pixels.
+    """
+    dist_inches = dist * 12
+    return 1.0 * dist_inches * width / obj_width
+
+
+def convert_to_network(x: np.ndarray, net_size: Tuple[int, int], clip_min: float, clip_max: float) -> np.ndarray:
+    """
+    Convert image to network format.
+    :param x: Input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:76]:
```suggestion

    :param x: Input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 36]
```diff
@@ -0,0 +1,671 @@
...
+    net_size: Tuple[int, int],
+    obj_width: float,
+    focal: float,
+    clip_min: float,
+    clip_max: float,
+    pts: Optional[np.ndarray] = None,
+) -> np.ndarray:
+    """
+    Apply transformation to input image.
+    :param x: Input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:112]:
```suggestion

    :param x: Input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 37]
```diff
@@ -0,0 +1,671 @@
...
+    crop_percent_range: Tuple[float, float],
+    off_x_range: Tuple[float, float],
+    off_y_range: Tuple[float, float],
+    blur_kernels: Union[Tuple[int, int], List[int]],
+    obj_width: float,
+    focal: float,
+) -> List[Tuple[float, float, float, int, float, float, float, float, float]]:
+    """
+    Sample transformation params.
+    :param num_xforms: The number of transforms to sample.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:176]:
```suggestion

    :param num_xforms: The number of transforms to sample.
```

File: art/attacks/evasion/graphite/utils.py [hunk 38]
```diff
@@ -0,0 +1,671 @@
...
+def add_noise(
+    x: np.ndarray,
+    mask: np.ndarray,
+    lbd: float,
+    theta: np.ndarray,
+    clip: bool = True,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Combines the image and noise to create a perturbed image.
+    :param x: Input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:216]:
```suggestion

    :param x: Input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 39]
```diff
@@ -0,0 +1,671 @@
...
+    lbd: float,
+    theta: np.ndarray,
+    net_size: Tuple[int, int],
+    clip_min: float,
+    clip_max: float,
+    pts: Optional[np.ndarray] = None,
+) -> List[np.ndarray]:
+    """
+    Get transformed images.
+    :param x: Input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:255]:
```suggestion

    :param x: Input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 40]
```diff
@@ -0,0 +1,671 @@
...
+    mask: "torch.Tensor",
+    xform: Tuple[float, float, float, int, float, float, float, float, float],
+    pts: Optional[np.ndarray],
+    net_size: Tuple[int, int],
+    clip_min: float,
+    clip_max: float,
+) -> "torch.Tensor":
+    """
+    Get transformed image, white-box setting.
+    :param x: Original input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:310]:
```suggestion

    :param x: Original input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 41]
```diff
@@ -0,0 +1,671 @@
...
+
+    return convert_to_network_wb(x_adv, net_size, clip_min, clip_max)
+
+
+def convert_to_network_wb(
+    x: "torch.Tensor", net_size: Tuple[int, int], clip_min: float, clip_max: float
+) -> "torch.Tensor":
+    """
+    Convert image to network format.
+    :param x: Input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:370]:
```suggestion

    :param x: Input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 42]
```diff
@@ -0,0 +1,671 @@
...
+    focal: float,
+    dist: float,
+    crop_percent: float,
+    crop_off_x: float,
+    crop_off_y: float,
+    pts: Optional[np.ndarray] = None,
+) -> np.ndarray:
+    """
+    Computes parameters for perspective transform for blackbox attack.
+    :param img: Input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:400]:
```suggestion

    :param img: Input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 43]
```diff
@@ -0,0 +1,671 @@
...
+    focal: float,
+    dist: float,
+    crop_percent: float,
+    crop_off_x: float,
+    crop_off_y: float,
+    pts: Optional[np.ndarray] = None,
+) -> "torch.Tensor":
+    """
+    Computes perspective transform for whitebox attack.
+    :param img: Input image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:435]:
```suggestion

    :param img: Input image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 44]
```diff
@@ -0,0 +1,671 @@
...
+    focal: float,
+    dist: float,
+    crop_percent: float,
+    crop_off_x: float,
+    crop_off_y: float,
+    pts: Optional[np.ndarray] = None,
+) -> Tuple[np.ndarray, int, int]:
+    """
+    Computes parameters for perspective transform.
+    :param angle: Angle to rotate.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:476]:
```suggestion

    :param angle: Angle to rotate.
```

File: art/attacks/evasion/graphite/utils.py [hunk 45]
```diff
@@ -0,0 +1,671 @@
...
+    h_mat: np.ndarray,
+    crop_percent: float,
+    crop_off_x: float,
+    crop_off_y: float,
+    ratio: float,
+    pts: Optional[np.ndarray] = None,
+) -> Tuple[float, float, float]:
+    """
+    Compute offsets and crop size for perspective transform.
+    :param w: Width of image.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:541]:
```suggestion

    :param w: Width of image.
```

File: art/attacks/evasion/graphite/utils.py [hunk 46]
```diff
@@ -0,0 +1,671 @@
...
+def run_predictions(
+    estimator: "CLASSIFIER_NEURALNETWORK_TYPE",
+    imgs: List[np.ndarray],
+    target: int,
+    batch_size: int,
+    err_rate: bool = True,
+) -> float:
+    """
+    Run model predictions over batch of input.
+    :param estimator: The model.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:623]:
```suggestion

    :param estimator: The model.
```

File: art/attacks/evasion/graphite/utils.py [hunk 47]
```diff
@@ -0,0 +1,671 @@
...
+            count = 0
+
+    return_val = (1.0 - num_successes * 1.0 / len(imgs)) if err_rate else num_successes * 1.0 / len(imgs)
+    return return_val
+
+
+def score_fn(mask: np.ndarray, tr_err: float, object_size: float, threshold: float = 0.75, lbd: float = 5):
+    """
+    Mask scoring function.
+    :param mask: The mask.
```
Comment (keykholt) [inline art/attacks/evasion/graphite/utils.py:661]:
```suggestion

    :param mask: The mask.
```

**Comment by beat-buesser** [comment]
Closing this PR because work continues in #1828

**Pull request closed without merging**